ngraph/core/reference/src/runtime/reference/fft.cpp

```diff
@@ -0,0 +1,533 @@
+                    if (is_power_of_two(length))
+                    {
+                        optimized_fft1d(length, fft_offset, stride, data, buffer, fft_kind);
+                    }
+                    else
+                    {
+                        naive_fft1d(length, fft_offset, stride, data, buffer, fft_kind);
+                    }
+                }
+            }
+
+            void fft(const float* input_data,
+                     const Shape& input_data_shape,
+                     const int64_t* axes_data,
+                     const Shape& axes_data_shape,
+                     float* fft_result,
+                     const Shape& output_shape,
+                     FFTKind fft_kind)
+            {
+                const complex_type* complex_input_data_ptr =
+                    reinterpret_cast<const complex_type*>(input_data);
+                complex_type* complex_output_ptr = reinterpret_cast<complex_type*>(fft_result);
+
+                const int64_t complex_data_rank = static_cast<int64_t>(input_data_shape.size() - 1);
+
+                const auto reversed_output_shape = reverse_shape(output_shape);
+                auto fft_axes = get_axes(axes_data, axes_data_shape, complex_data_rank);
+                reverse_fft_axes(fft_axes, complex_data_rank);
+
+                const int64_t fft_rank = fft_axes.size();
+                const auto fft_lengths = get_lengths(reversed_output_shape, fft_axes);
+                const auto fft_strides = compute_strides(fft_lengths);
+                const int64_t fft_size = fft_strides[fft_rank];
+
+                if (fft_size <= 0)
+                {
+                    return;
+                }
+
+                std::vector<complex_type> data(fft_size);
+
+                const auto outer_axes = get_outer_axes(fft_axes, complex_data_rank);
+                const int64_t outer_rank = outer_axes.size();
+                const auto outer_lengths = get_lengths(reversed_output_shape, outer_axes);
+                const auto outer_strides = compute_strides(outer_lengths);
+                const int64_t outer_size = outer_strides[outer_rank];
+
+                int64_t buffer_size = compute_buffer_size(fft_lengths);
+                std::vector<complex_type> buffer(buffer_size);
+
+                const auto output_strides = compute_strides(reversed_output_shape);
+                const auto output_fft_strides = get_lengths(output_strides, fft_axes);
+                const auto output_outer_strides = get_lengths(output_strides, outer_axes);
+                const auto reversed_input_shape = reverse_shape(input_data_shape);
+                const auto input_fft_lengths = get_lengths(reversed_input_shape, fft_axes);
+                const auto input_strides = compute_strides(reversed_input_shape);
+                const auto input_fft_strides = get_lengths(input_strides, fft_axes);
+                const auto input_outer_strides = get_lengths(input_strides, outer_axes);
```

Comment:
Maybe this initialization can be moved to a separate function and applied to input_data_shape and output_shape?

Reply:
Done.